src/devices/src/virtio/net/device.rs

```diff
@@ -302,94 +336,177 @@ impl Net {
         success
     }
 
-    // Copies a single frame from `self.rx_frame_buf` into the guest.
-    fn do_write_frame_to_guest(&mut self) -> std::result::Result<(), FrontendError> {
-        let mut result: std::result::Result<(), FrontendError> = Ok(());
+    // Reserves available writable buffers from the RX queue.
+    // The method will try to reserve as many buffers as needed to sucesfully
+    // store the packet that is in the device receive buffer.
+    fn reserve_guest_wbuffers(
+        &mut self,
+        buffs: &mut Vec<(u16, Vec<(GuestAddress, u32)>)>
+    ) -> std::result::Result<(), FrontendError> {
         let mem = match self.device_state {
             DeviceState::Activated(ref mem) => mem,
             // This should never happen, it's been already validated in the event handler.
             DeviceState::Inactive => unreachable!(),
         };
 
+        let use_mrg = self.use_mrg_rx_buffs();
         let queue = &mut self.queues[RX_INDEX];
-        let head_descriptor = queue.pop(mem).ok_or_else(|| {
-            METRICS.net.no_rx_avail_buffer.inc();
-            FrontendError::EmptyQueue
-        })?;
-        let head_index = head_descriptor.index;
+        let mut total_reserved_sz: usize = 0;
+
+        // When not using mergeable RX buffers, the device must add the packet into
+        // a single descriptor chain.
+        let limit = if use_mrg { queue.actual_size() } else { 1 };
+        let mut tries = 0;
+        while total_reserved_sz < self.rx_bytes_read && tries < limit {
+            let head_descriptor = queue.pop(mem).ok_or_else(|| {
+                METRICS.net.no_rx_avail_buffer.inc();
+                if !buffs.is_empty() {
+                    // Found some descriptors but total size is insufficient.
+                    // Don't use these descriptors and wait for the driver to
+                    // add more.
+                    queue.undo_multiple_pop(buffs.len() as u16);
+                    buffs.clear();
+                    FrontendError::InsufficientBuffers
+                } else {
+                    FrontendError::EmptyQueue
+                }
+            })?;
+
+            // When using mergeable RX buffers, the driver usually adds one descriptor per chain.
+            // When using big buffers the driver must add MAX_BUFFER_SIZE / PAGE_SIZE buffers per chain.
+            let mut desc_chain_wbuffs = if use_mrg { Vec::with_capacity(1) } else { Vec::with_capacity(17) };
```

Comment:
let's define a constant here derived from `MAX_BUFFER_SIZE`:
```suggestion
            let mut desc_chain_wbuffs = if use_mrg { Vec::with_capacity(1) } else { Vec::with_capacity(MAX_BUFFER_PAGES) };
```